uvicorn/lifespan/on.py

```diff
@@ -22,12 +22,14 @@ def __init__(self, config: Config) -> None:
         self.startup_failed = False
         self.shutdown_failed = False
         self.should_exit = False
+        self.main_task = None
 
     async def startup(self) -> None:
         self.logger.info("Waiting for application startup.")
 
         loop = asyncio.get_event_loop()
-        loop.create_task(self.main())
+        task = loop.create_task(self.main())
+        self.main_task = task  # Keep a hard reference to prevent garbage collection
```

Comment:
```suggestion
                main_lifespan_task = loop.create_task(self.main())  # noqa: F841  # Keep a hard reference to prevent garbage collection, see https://github.com/encode/uvicorn/pull/972
```

Reply:
I think this is enough just to create the variable, let's also add a comment pointing to the PR as it's definitely non-trivial

Reply:
Heads up, I realized recently that this isn't enough to just create the variable. If the app waits on any io immediately after sending the startup complete message it will still get collected by GC. Probably not something many apps do though so relatively low impact.